src/user-support.c

```diff
@@ -17,49 +17,120 @@
 #include "config.h"
 #include "user-support.h"
 
-#include <errno.h>
 #include <stdlib.h>
 #include <sys/stat.h>
+#include <unistd.h>
 
 #include "utils.h"
 
-void setup_user_data()
+void setup_user_data(uid_t uid, gid_t gid)
 {
 	const char *user_data = getenv("SNAP_USER_DATA");
-
 	if (user_data == NULL)
 		return;
-
 	// Only support absolute paths.
 	if (user_data[0] != '/') {
 		die("user data directory must be an absolute path");
 	}
-
+	// Describe how the created directory should look like.
+	int hint_SNAP_USER_DATA(int depth, const char *segment,
+				mode_t * mode_p, uid_t * uid_p, gid_t * gid_p) {
+		switch (depth) {
+		case 0:
+			// That's /
+			break;
+		case 1:
+			// that's /home
+			break;
+		case 2:
+			/// that's /home/$LOGNAME
+			break;
+		case 3:
+			// that's /home/$LOGNAME/snap
+			// NOTE: fall-through
+		case 4:
+			// that's /home/$LOGNAME/snap/$SNAP_NAME
+			// NOTE: fall-through
+		case 5:
+			// that's /home/$LOGNAME/snap/$SNAP_NAME/$SNAP_REVISION
+			*uid_p = uid;
+			*gid_p = gid;
+			*mode_p = 0755;
+			break;
+		default:
+			die("unexpected SNAP_USER_DATA segment: %s at depth %d",
+			    segment, depth);
+		}
+		return 0;
+	}
```

Comment:
This is pretty complicated and is assuming a particular directory hierarchy. People might have their HOME in /srv, /var, etc (they'd have to adjust /etc/apparmor.d/tunables/home* for this, but it is at least runtime tunable as opposed to this).